# Cross-Validating across Time using Multiple Linear Regression

A major challenge of modeling data that is collected over time is that there is likely to be correlation between consecutive entries. This poses problems because a key assumption that is made when fitting a linear regression model is that the observations are independent of one another. In order to work around this, we can instead evaluate the model by taking the first subset of observations (ordered by time), forecasting the next observation and recording the mean squared error. We can repeat this process, obtain the sum of the mean squared errors, and now we have a metric by which we can "cross-validate" the model!

We will write an algorithm to demonstrate this process using Python. Data for this particular example is air quality data from the UC Irvine Machine Learning Repository. The observations were collected between March 2004 to February 2005 and data was collected using metal oxide chemical sensors that were embedded in a device located in a significantly polluted area in Italy. More information on the dataset is available [here](https://archive.ics.uci.edu/dataset/360/air+quality).

We'll go ahead and read in the data now:

In [181]:
#Import necessary modules
import pandas as pd
import numpy as np
from sklearn import linear_model

# Read in the dataset from the UCI Machine Learning Repository library
!pip install ucimlrepo
import ucimlrepo as uci

air_quality = uci.fetch_ucirepo(id=360)
aq = air_quality.data.features

#Sets the Date column to a datetime object
#This ensures the final data is in chronological order
#https://pandas.pydata.org/docs/user_guide/timeseries.html
aq['Date_DT'] = aq.Date.astype('datetime64[ns]')

aq.head()

Date      Time  CO(GT)  PT08.S1(CO)  NMHC(GT)  C6H6(GT)  \
0  3/10/2004  18:00:00     2.6         1360       150      11.9   
1  3/10/2004  19:00:00     2.0         1292       112       9.4   
2  3/10/2004  20:00:00     2.2         1402        88       9.0   
3  3/10/2004  21:00:00     2.2         1376        80       9.2   
4  3/10/2004  22:00:00     1.6         1272        51       6.5   

   PT08.S2(NMHC)  NOx(GT)  PT08.S3(NOx)  NO2(GT)  PT08.S4(NO2)  PT08.S5(O3)  \
0           1046      166          1056      113          1692         1268   
1            955      103          1174       92          1559          972   
2            939      131          1140      114          1555         1074   
3            948      172          1092      122          1584         1203   
4            836      131          1205      116          1490         1110   

      T    RH      AH    Date_DT  
0  13.6  48.9  0.7578 2004-03-10  
1  13.3  47.7  0.7255 2004-03-10  
2  11.9  54.0  0.7502 2004-03-10  
3  11.0  60.0  0.7867 2004-03-10  
4  11.2  59.6  0.7888 2004-03-10

Before the models can be evaluated, we will need to clean up this data a little bit. First, there are missing values that we need to get rid of in some of the parameters we will be working with: `C6H6(GT)`, `CO(GT)`, `T`, `RH`, and `AH`.

In [59]:
#Clear out missing values (corresponding to -200 in this dataset)
aq_no_missing = aq[(aq['C6H6(GT)'] != -200) & (aq['CO(GT)'] != -200) & (aq['T'] != -200) & (aq['RH'] != -200) & (aq['AH'] != -200)]

 Next, the values we wish to work with are the average of each parameter of interest across a given `Date`, so we will group the data by `Date` and return the average for each of the five aforementioned variables. Note that we expect 347 observations after completing this operation.

In [60]:
#Group data by date, return the average across each day
aq_avg = aq_no_missing.groupby('Date_DT')[['C6H6(GT)', 'CO(GT)','T','RH','AH']].mean()

#Append a date column so the data is a bit easier to iterate over
aq_avg['Day'] = range(1, len(aq_avg) + 1)
aq_avg

C6H6(GT)    CO(GT)          T         RH        AH  Day
Date_DT                                                             
2004-03-10   8.450000  1.966667  12.033333  54.900000  0.765633    1
2004-03-11   8.269565  2.239130   9.826087  64.230435  0.777039    2
2004-03-12  12.177273  2.804545  11.618182  50.190909  0.665164    3
2004-03-13  11.121739  2.695652  13.121739  50.682609  0.733013    4
2004-03-14   9.830435  2.469565  16.182609  48.317391  0.849209    5
...               ...       ...        ...        ...       ...  ...
2005-03-31   5.220833  1.387500  17.550000  50.083333  0.951917  343
2005-04-01   3.526087  1.108696  16.026087  35.404348  0.631135  344
2005-04-02   2.529167  0.854167  15.483333  32.225000  0.546167  345
2005-04-03   4.316667  1.141667  18.383333  33.695833  0.617583  346
2005-04-04   8.985714  2.078571  17.328571  41.842857  0.721250  347

[347 rows x 6 columns]

Now, we are ready to start modeling!

## The Models

In this context, we are interested in predicting/inferring the concentration of Benzene (`C6H6(GT)`). We have two models in mind:

- A simple linear regression (SLR) model using the concentration of CO `CO(GT)` as the predictor.
- A multiple linear regression (MLR) model using the concentration of CO `CO(GT)` as well as the temperature, relative humidity, and absolute humidity (`T`, `RH`, `AH`) as predictors.

Let's fit the basic models now:

In [54]:
#Fit the SLR model
reg = linear_model.LinearRegression()
reg.fit(X = aq_avg['CO(GT)'].values.reshape(-1,1), y = aq_avg['C6H6(GT)'])

string1 = "Intercept: {}"
string2 = "Coefficient(s): {}"
print(string1.format(reg.intercept_))
print(string2.format(reg.coef_))

Intercept: 0.644770948344199
Coefficient(s): [4.56536116]


In [57]:
#Fit the MLR model
reg.fit(X = aq_avg[['CO(GT)','T','RH','AH']], y = aq_avg['C6H6(GT)'])

string1 = "Intercept: {}"
string2 = "Coefficient(s): {}"
print(string1.format(reg.intercept_))
print(string2.format(reg.coef_))

Intercept: -1.8377694729981755
Coefficient(s): [ 4.77080433  0.11973259 -0.01620259  0.68866811]


Keep in mind that these models are "naive" in a sense, since they do not take the date into account and the assumption that observations are independent is currently in question. But now that we know how these models are going to be fit, we can write the functions that will evaluate them sequentially.

## Creating Functions to Sequentially Validate Linear Regression Models

The general algorithm for this is as follows:

- Fit the model using the observations from the first `n` units of time (in our context, days), where `n` is a user-specified number.
- Predict the response recorded at the `n + 1` unit of time. Get the mean squared error (MSE) from this prediction.
- Fit the model using the first `n + 1`, and get the MSE for a prediction for the observation at `n + 2`.
- Repeat until the model is fit to the full dataset.
- Get a sum of the MSE values obtained across all iterations of this process, which is used as a metric to evaluate the model. The lower the MSE value, the better.

This particular implementation will rely on two functions:

- A "helper" function that is given a subset of the data, and returns the MSE of the model that is fit to that subset of the data.
- A function that handles the iteration across different repetitions of the algorithm and evaluates the sum of the MSE values collected over time.

We'll start with the helper:

In [205]:
# Define a function to evaluate a subset of a model, predict the next chronological value, and return the MSE
def getCurrentMSE(X, y, day):
  """
  Accepts the following as arguments:
  X = a 1D or 2D object, used as predictors for a linear regression model. Should be of length `day + 1`.
  y = a 1D object, used as the response variable for a linear regression model. Should be of length `day + 1`.
  day = an integer, indicates the number of days to be used as training data in this iteration.
  Returns a float, representing the mean squared error (MSE) of this iteration's model fit according to the above arguments.
  """
  #Get the training data
  X_train = X.iloc[:-1]
  X_test = X.iloc[day]
  y_train = y.iloc[:-1]
  y_test = y.iloc[day]

  # Check if the number of predictors == 1, reshape data as needed, get raw arrays
  if type(X) == pd.core.series.Series:
    X_train = X_train.values.reshape(-1, 1)
    y_train = y_train.values
    X_test = np.array(X_test).reshape(-1, 1) #sklearn.linear_model.predict() expects a 2D array, which is accomplished here
  else:
    X_train = X_train.values
    y_train = y_train.values
    X_test = X_test.values.reshape(-1, X_test.shape[0])


  #Fit the model
  reg = linear_model.LinearRegression()
  reg.fit(X = X_train, y = y_train)

  #Predict the test observation
  y_hat = reg.predict(X_test)[0] #returns a 1x1 array, calling [0] to get the value inside

  #Calculate MSE and return
  return((y_test - y_hat)**2)

#Test the function
getCurrentMSE(X = aq_avg.loc[aq_avg['Day'] <= 251,['CO(GT)','T','RH','AH']], y = aq_avg.loc[aq_avg['Day'] <= 251,'C6H6(GT)'], day = 250)

np.float64(1.632153205961474)

Looks like everything is in order, as the test run of the function evaluated the MSE of the 250-day subset of the initial data to be approximately 1.632, which is a reasonable value. Next, we can write the main function:

In [206]:
#Define a function to iterate through repititions of the findMSE function, and return a "full" MSE to act as a model metric
def getMSE(X, y, day = 250):
  """
  Accepts the following as arguments:
  X = a 1D or 2D object, used as predictors for a linear regression model. Should be of a length greater than the `day` argument.
  y = a 1D object, used as the response variable for a linear regression model. Should match the length of the `X` argument.
  day = an integer, representing the last observation that should be used as the training data. Default is 250.
  Returns a float, representing the mean squared error (MSE) of the overall model fit according to the above arguments.
  """
  #Initialize MSE to 0
  mse = 0

  #Iterate through the sequential subsets and add to the current MSE
  for i in range(day, len(X)):
    idx = i + 1
    mse += getCurrentMSE(X.iloc[:idx], y.iloc[:i+1], day = i)

  return(mse)

#Test function
getMSE(X = aq_avg.loc[:,['RH','AH']], y = aq_avg.loc[:,'C6H6(GT)'], day = 250)

np.float64(1450.6513814451976)

We've returned a singular value that appears reasonable for the parameters we tested. This indicates that the function works as intended!

Finally, we can run this algorithm on the two models we proposed earlier and evaluate which one is "best" for estimating the concentration of benzene.

In [207]:
#Evaluate the MSE of the models and print the output
slr_mse = "Mean Squared Error of the SLR model: {}"
mlr_mse = "Mean Squared Error of the MLR model: {}"
print(slr_mse.format(getMSE(X = aq_avg['CO(GT)'], y = aq_avg['C6H6(GT)'], day = 250)))
print(mlr_mse.format(getMSE(X = aq_avg[['CO(GT)','T','RH','AH']], y = aq_avg['C6H6(GT)'], day = 250)))


Mean Squared Error of the SLR model: 718.0810286685651
Mean Squared Error of the MLR model: 494.34222176464584


With that, it appears that the MLR model more accurately predicted future values of benzene concentration, since the mean squared error is lower for the MLR model compared to the SLR model. We'll fit the full dataset to the MLR model as a result.

In [208]:
reg.fit(X = aq_avg[['CO(GT)','T','RH','AH']], y = aq_avg['C6H6(GT)'])

string1 = "Intercept: {}"
string2 = "Coefficient(s): {}"
print(string1.format(reg.intercept_))
print(string2.format(reg.coef_))

Intercept: -1.8377694729981755
Coefficient(s): [ 4.77080433  0.11973259 -0.01620259  0.68866811]


The fitted model is as follows:

`Estimated C6H6(GT)` = -1.84 + (4.77 * `CO(GT)`) + (0.12 * `T`) - (0.016 * `RH`) + (0.689 * `AH`)